---
title: "Assignment 1"
output:
  html_document:
    df_print: paged
---
### Kuai Yu; 1004801912

# Data set
Geo Accession id: GSE152323

Title: Gene expression analysis of RECQ1-regulated transcriptome in breast 
cancer cells.


### Q1. What are the control and test conditions of the dataset?
In this test, the controls are MCF7 cells transfected with control siRNA and
the tests are MCF7 cells transfected with ESR1 siRNAs (SmartPool) and RECQ1
siRNAs (SmartPool) for 48 hours.

### Q2. Why is the dataset of interest to you?
This paper is quite new in topics relating to breast cancer, so it might provide
some new insights in this disease.

### Q3. Were there expression values that were not unique for specific genes? How did you handle these?
Based on the summarized_gene_counts later, there is no duplicated gene

### Q4. Were there expression values that could not be mapped to current HUGO symbols?
Yes, there exits expression values that could not be mapped to HUGO symbols. 
There are in total 718 expression values are NA or "".

### Q5. How many outliers were removed?
I do not remove any outlier.

### Q6. How did you handle replicates?
There is no replicates for my set of data.


### Q7. What is the final coverage of your dataset?
13526 out of 56632.

## Load essential packages

```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

if (!requireNamespace("GEOquery", quietly = TRUE))
  BiocManager::install("GEOquery")

if (!requireNamespace("edgeR", quietly = TRUE))
  BiocManager::install("edgeR")

if (!requireNamespace("biomaRt", quietly = TRUE))
  BiocManager::install("biomaRt")

if (!requireNamespace("knitr", quietly = TRUE))
  install.packages("knitr")
```


```{r}
library(GEOmetadb)
library(edgeR)
library(biomaRt)
library(knitr)
```

# Download the data
```{r}
geo_accession <- "GSE152323"
sfiles <- NULL

# Checking if the file already exists
if (! file.exists("./GSE152323/")) {
  sfiles <- getGEOSuppFiles(geo_accession)
  saveRDS(sfiles, "./sfiles.rds")
} else {
  sfiles <- readRDS("./sfiles.rds")
}


fnames = rownames(sfiles)

# File is separated by \t
recq1_exp = read.delim(fnames[1], 
                         header = TRUE, 
                         check.names = FALSE, 
                         sep = "\t")

kable(head(recq1_exp), type = "html")
```

```{r}
dim(recq1_exp)

colnames(recq1_exp)

# Missing first column name
colnames(recq1_exp)[1] <- "ensembl_id" 

# Also names of samples are too long
raw_samples_names <- unname(
    unlist(data.frame(strsplit(colnames(recq1_exp)[2: length(colnames(recq1_exp))], 
                               "/")))[c(TRUE, FALSE)])

colnames(recq1_exp)[2: ncol(recq1_exp)] <- raw_samples_names

colnames(recq1_exp)


```

## Filter low counts

```{r}
cpms = cpm(recq1_exp[, 2: ncol(recq1_exp)])
rownames(cpms) <- recq1_exp[, 1]

# there are 6 samples
keep = rowSums(cpms > 1) >= 6
recq1_exp_filtered = recq1_exp[keep, ]

dim(recq1_exp_filtered)


```

# Assess distribution of data


## Box plot
```{r}
data2plot <- log2(cpm(recq1_exp_filtered[, 2: ncol(recq1_exp_filtered)]))
boxplot(data2plot, 
        xlab = "Samples", 
        ylab = "log2 CPM", 
        las = 2, 
        cex = 0.3, 
        cex.lab = 0.3,
        cex.axis = 0.5, 
        main = "recq1 RNASeq Samples")

```

## Density Plot

```{r}
counts_density <- 
  apply(log2(cpm(recq1_exp_filtered[, 2:length(colnames(recq1_exp_filtered))])),
        2, density)
 #calculate the limits across all the samples
xlim <- 0; ylim <- 0
for (i in 1:length(counts_density)) {
  xlim <- range(c(xlim, counts_density[[i]]$x)); 
  ylim <- range(c(ylim, counts_density[[i]]$y))
}
 cols <- rainbow(length(counts_density))
 ltys <- rep(1, length(counts_density))
 
 #plot the first density plot to initialize the plot
 plot(counts_density[[1]], 
      xlim=xlim, 
      ylim=ylim, 
      type="n", 
      ylab="Smoothing density of log2-CPM", 
      main="Density Plot before Normalization", 
      cex.lab = 0.8)
 
 #plot each line
 for (i in 1:length(counts_density)) {
   lines(counts_density[[i]], 
         col = cols[i], 
         lty = ltys[i])
 }
   
 #create legend
 legend("topright", colnames(data2plot), 
 col=cols, lty=ltys, cex=0.6, 
 border ="blue", text.col = "green4", 
 merge = TRUE, bg = "gray90")
```



## Define groups
```{r}
# Group them based on different cell types
cell_type <- unlist(strsplit(raw_samples_names, "_"))[c(FALSE, TRUE, FALSE)]
samples <- data.frame(cell_type = cell_type)
rownames(samples) <- colnames(recq1_exp[2: length(recq1_exp)])

kable(samples, format = "html")
```




## Checking duplicated genes
```{r}
summarized_gene_counts <- sort(table(recq1_exp$ensembl_id), decreasing = TRUE)
kable(head(summarized_gene_counts), format = "html")

# There is no dupliacted genes
```


# Map

```{r}
httr::set_config(httr::config(ssl_verifypeer = FALSE)) # sometimes code bellow runs error, this line solves it
ensembl <- useMart("ensembl")
datasets <- listDatasets(ensembl)
ensembl <- useDataset("hsapiens_gene_ensembl", mart = ensembl)
```


```{r}
conversion_stash <- "./recq1_id_conversion.rds"
if (file.exists(conversion_stash)) {
  recq1_id_conversion <- readRDS(conversion_stash)
} else {
  recq1_id_conversion <- getBM(attributes = c("ensembl_gene_id", "hgnc_symbol"),
                                 filters = c("ensembl_gene_id"),
                                 values = recq1_exp_filtered$ensembl_id,
                                 mart = ensembl)
  saveRDS(recq1_id_conversion, conversion_stash)
}
```




## Merge new identifiers

```{r}
recq1_exp_filtered_annot <- merge(recq1_id_conversion, 
                                 recq1_exp_filtered, 
                                 by.x = 1, 
                                 by.y = 1, 
                                 all.y=TRUE)
kable(recq1_exp_filtered_annot[1:10, ], type = "html")
```

## Check and deal with missing identifiers

```{r}
missing_index <- which(is.na(recq1_exp_filtered_annot$hgnc_symbol) | 
                         recq1_exp_filtered_annot$hgnc_symbol == "")
num_missing <- sum(is.na(recq1_exp_filtered_annot$hgnc_symbol) | 
                         recq1_exp_filtered_annot$hgnc_symbol == "")

num_missing # Print number of missing identifier

# After checking most of genes, they are mostly deprecated or curated; thus, I use
# their gene id as their name since they might be useful.

ensembl_id_missing_gene <- recq1_exp_filtered_annot[missing_index, ]
recq1_exp_filtered_annot$hgnc_symbol[missing_index] <- 
  recq1_exp_filtered_annot$ensembl_gene_id[missing_index]

kable(recq1_exp_filtered_annot[missing_index[1: 10], ], type = "html")
```



## Check duplicated mapping
```{r}
summarized_mapping_counts <- sort(table(recq1_exp_filtered_annot$hgnc_symbol), 
                                  decreasing = TRUE)
head(summarized_mapping_counts)
summarized_mapping_counts_2 <- sort(table(
  recq1_exp_filtered_annot$ensembl_gene_id), decreasing = TRUE)
head(summarized_mapping_counts_2)

# There is no duplicated mapping in either way
```

### Applying Normalization (TMM)

```{r}
filtered_data_matrix <- as.matrix(recq1_exp_filtered_annot[,3: ncol(recq1_exp_filtered_annot)])
rownames(filtered_data_matrix) <- recq1_exp_filtered_annot$hgnc_symbol
d = DGEList(counts = filtered_data_matrix, group = samples$cell_type)
d = calcNormFactors(d) 
normalized_counts <- cpm(d)
kable(normalized_counts[1: 10, ], type = "html")
nrow(normalized_counts)
```

### MDS plot
```{r}
plotMDS(d, labels=rownames(samples),
 col = c("darkgreen","blue")[factor(samples$cell_type)], cex = 0.6)

```



```{r}
# density plot after normalized by defined groups
counts_density <- apply(log2(normalized_counts), 2, density)

xlim <- 0; ylim <- 0
for (i in 1:length(counts_density)) {
  xlim <- range(c(xlim, counts_density[[i]]$x)); 
  ylim <- range(c(ylim, counts_density[[i]]$y))
}
 cols <- rainbow(length(counts_density))
 ltys <- rep(1, length(counts_density))
 
 #plot the first density plot to initialize the plot
 plot(counts_density[[1]], 
      xlim=xlim, 
      ylim=ylim, 
      type="n", 
      ylab="Smoothing density of log2-CPM", 
      main="Density Plot after Normalization", 
      cex.lab = 0.8)
 #plot each line
 for (i in 1:length(counts_density)) {
   lines(counts_density[[i]], 
         col = cols[i], 
         lty = ltys[i])
 }
   
 #create legend
 legend("topright", colnames(data2plot), 
        col=cols, lty=ltys, cex=0.5, 
        border ="blue", text.col = "green4", 
        merge = TRUE, bg = "gray90")
```
### Dispersion
```{r}
model_design <- model.matrix(~samples$cell_type+0)
d <- estimateDisp(d, model_design)
```

```{r}
plotBCV(d,col.tagwise = "black",col.common = "red")

```
```{r}
plotMeanVar(d, show.raw.vars = TRUE, 
            show.tagwise.vars=TRUE, 
            show.ave.raw.vars = TRUE, 
            NBline=TRUE,
            show.binned.common.disp.vars = TRUE)
```
 
# Final Result
```{r}
final_counts <- data.frame(ensembl_gene_id = recq1_exp_filtered_annot$ensembl_gene_id)
final_counts <- cbind(final_counts, as.data.frame(normalized_counts))
kable(final_counts[1: 10, ], type = "html")
write.csv(final_counts, file = "./final_counts.csv")
```

# Reference
Li, X.L., Lu, X., Parvathaneni, S., Bilke, S., Zhang, H., Thangavel, S., Vindigni, A., Hara, T., Zhu, Y., Meltzer, P.S., Lal, A., Sharma, S., 2014. Identification of RECQ1-regulated transcriptome uncovers a role of RECQ1 in regulation of cancer cell migration and invasion. Cell Cycle 13, 2431–2445.. doi:10.4161/cc.29419

Steipe, B. (n.d.). GEO2R - “A B C”. from
https://bcb420-2022.github.io/Bioinfo_Basics/de-genes.html#geo2r.

Robinson MD, McCarthy DJ and Smyth GK , 2010. edgeR: a Bioconductor package for differential expression analysis of digital gene expression data.
  Bioinformatics 26, 139-140

Davis, S. and Meltzer, P. S, 2007, GEOquery: a bridge between the Gene Expression Omnibus (GEO) and BioConductor. Bioinformatics, 14, 1846-1847

Steffen Durinck, Paul T. Spellman, Ewan Birney and
  Wolfgang Huber. 2009 Mapping identifiers for the integration of genomic datasets with the R/Bioconductor package biomaRt. , Nature Protocols 4, 1184-1191.
